### PR TITLE
vweb: host attribute

### PR DESCRIPTION
**Pull request by Lenni0451**
Add support for a [host: 'example.org'] attribute in vweb.
This allows having host specific endpoints when using multiple domains (subdomains) for a page.

### <samp>🤖 Generated by Copilot at e14dfa7</samp>

This pull request adds a new feature to the `vweb` module that allows defining host-based routes for web endpoints. It modifies the `vweb` module's code and documentation to support and explain this feature. It affects the files `vlib/vweb/parse.v`, `vlib/vweb/vweb.v`, and `vlib/vweb/README.md`.

### <samp>🤖 Generated by Copilot at e14dfa7</samp>

*  Add host-based routing feature to vweb module ([link](https://github.com/vlang/v/pull/18288/files?diff=unified&w=0#diff-825536522629231585c41f7252c76ecba1ad4acada7d07dfe8aae1678aad54afL7-R9), [link](https://github.com/vlang/v/pull/18288/files?diff=unified&w=0#diff-825536522629231585c41f7252c76ecba1ad4acada7d07dfe8aae1678aad54afR16), [link](https://github.com/vlang/v/pull/18288/files?diff=unified&w=0#diff-825536522629231585c41f7252c76ecba1ad4acada7d07dfe8aae1678aad54afR40-R44), [link](https://github.com/vlang/v/pull/18288/files?diff=unified&w=0#diff-825536522629231585c41f7252c76ecba1ad4acada7d07dfe8aae1678aad54afL52-R59), [link](https://github.com/vlang/v/pull/18288/files?diff=unified&w=0#diff-694cdbe16b700f23ab1ec25807adbb3864af7d46788aa84b10f725eb9a401c80R186), [link](https://github.com/vlang/v/pull/18288/files?diff=unified&w=0#diff-694cdbe16b700f23ab1ec25807adbb3864af7d46788aa84b10f725eb9a401c80L402-R403), [link](https://github.com/vlang/v/pull/18288/files?diff=unified&w=0#diff-694cdbe16b700f23ab1ec25807adbb3864af7d46788aa84b10f725eb9a401c80R411), [link](https://github.com/vlang/v/pull/18288/files?diff=unified&w=0#diff-694cdbe16b700f23ab1ec25807adbb3864af7d46788aa84b10f725eb9a401c80L415-R417), [link](https://github.com/vlang/v/pull/18288/files?diff=unified&w=0#diff-694cdbe16b700f23ab1ec25807adbb3864af7d46788aa84b10f725eb9a401c80L439-R446), [link](https://github.com/vlang/v/pull/18288/files?diff=unified&w=0#diff-694cdbe16b700f23ab1ec25807adbb3864af7d46788aa84b10f725eb9a401c80R625-R626), [link](https://github.com/vlang/v/pull/18288/files?diff=unified&w=0#diff-694cdbe16b700f23ab1ec25807adbb3864af7d46788aa84b10f725eb9a401c80L638-R642), [link](https://github.com/vlang/v/pull/18288/files?diff=unified&w=0#diff-694cdbe16b700f23ab1ec25807adbb3864af7d46788aa84b10f725eb9a401c80L645-R653), [link](https://github.com/vlang/v/pull/18288/files?diff=unified&w=0#diff-694cdbe16b700f23ab1ec25807adbb3864af7d46788aa84b10f725eb9a401c80L693-R741), [link](https://github.com/vlang/v/pull/18288/files?diff=unified&w=0#diff-694cdbe16b700f23ab1ec25807adbb3864af7d46788aa84b10f725eb9a401c80L722-R767))
  * Modify `parse_attrs` function in `vlib/vweb/parse.v` to return host name as fourth value ([link](https://github.com/vlang/v/pull/18288/files?diff=unified&w=0#diff-825536522629231585c41f7252c76ecba1ad4acada7d07dfe8aae1678aad54afL7-R9), [link](https://github.com/vlang/v/pull/18288/files?diff=unified&w=0#diff-825536522629231585c41f7252c76ecba1ad4acada7d07dfe8aae1678aad54afR16), [link](https://github.com/vlang/v/pull/18288/files?diff=unified&w=0#diff-825536522629231585c41f7252c76ecba1ad4acada7d07dfe8aae1678aad54afR40-R44), [link](https://github.com/vlang/v/pull/18288/files?diff=unified&w=0#diff-825536522629231585c41f7252c76ecba1ad4acada7d07dfe8aae1678aad54afL52-R59))
  * Add `host` field to `Route` struct and assign it in `vweb.init` method ([link](https://github.com/vlang/v/pull/18288/files?diff=unified&w=0#diff-694cdbe16b700f23ab1ec25807adbb3864af7d46788aa84b10f725eb9a401c80R186), [link](https://github.com/vlang/v/pull/18288/files?diff=unified&w=0#diff-694cdbe16b700f23ab1ec25807adbb3864af7d46788aa84b10f725eb9a401c80L402-R403), [link](https://github.com/vlang/v/pull/18288/files?diff=unified&w=0#diff-694cdbe16b700f23ab1ec25807adbb3864af7d46788aa84b10f725eb9a401c80R411))
  * Add `host` parameter to `ControllerHandler` type alias and pass it in `vweb.controller` and `vweb.handle_conn` methods ([link](https://github.com/vlang/v/pull/18288/files?diff=unified&w=0#diff-694cdbe16b700f23ab1ec25807adbb3864af7d46788aa84b10f725eb9a401c80L415-R417), [link](https://github.com/vlang/v/pull/18288/files?diff=unified&w=0#diff-694cdbe16b700f23ab1ec25807adbb3864af7d46788aa84b10f725eb9a401c80L439-R446), [link](https://github.com/vlang/v/pull/18288/files?diff=unified&w=0#diff-694cdbe16b700f23ab1ec25807adbb3864af7d46788aa84b10f725eb9a401c80L638-R642))
  * Add `host` parameter to `handle_route` function and check it against route host name before matching ([link](https://github.com/vlang/v/pull/18288/files?diff=unified&w=0#diff-694cdbe16b700f23ab1ec25807adbb3864af7d46788aa84b10f725eb9a401c80L645-R653), [link](https://github.com/vlang/v/pull/18288/files?diff=unified&w=0#diff-694cdbe16b700f23ab1ec25807adbb3864af7d46788aa84b10f725eb9a401c80L693-R741))
  * Extract host name from request `Host` header in `vweb.handle_conn` method ([link](https://github.com/vlang/v/pull/18288/files?diff=unified&w=0#diff-694cdbe16b700f23ab1ec25807adbb3864af7d46788aa84b10f725eb9a401c80R625-R626))
  * Add missing `return` statement after route handler call in `handle_route` function ([link](https://github.com/vlang/v/pull/18288/files?diff=unified&w=0#diff-694cdbe16b700f23ab1ec25807adbb3864af7d46788aa84b10f725eb9a401c80L722-R767))
* Update README.md file of vweb module to explain how to use `host` attribute ([link](https://github.com/vlang/v/pull/18288/files?diff=unified&w=0#diff-0363f38173fe973cfa73ea26ebe6ed582a470e86ad0ef0b249acde5755c22f80R254-R270))

### CODE REVIEW

**Comment by medvednikov** [comment]
Nice! :)

**Comment by Casper64** [comment]
Nice! Is it still possible to test on localhost?

**Comment by Casper64** [comment]
Maybe hosts could also be an option on the Controller struct? So it's possible for all routes on a controller to belong to a a specific hostname without having to type the attribute

**Comment by Lenni0451** [comment]
> Nice! Is it still possible to test the routes with the host attribute on localhost?

If you have a host set it's not possible because of duplicate routes. The best way would be to temporarily add the host to your hosts file.

**Comment by Lenni0451** [comment]
> Maybe hosts could also be an option on the Controller struct? So it's possible for all routes on a controller to belong to a a specific hostname without having to type the attribute

I have not yet used controllers but it sounds like a good idea. I have to see if I get it to work.

**Comment by Lenni0451** [comment]
I have to duplicate the the `controller` function to accept an additional url but it should be no problem to also add support for that.

**Comment by Lenni0451** [comment]
I added host support to controllers. Should I push it with this PR or should I make a new one as soon as this gets accepted?

**Comment by medvednikov** [comment]
@Casper64 do you think this one can be merged as is?

**Comment by Casper64** [comment]
> @Casper64 do you think this one can be merged as is?

Yes should be fine

**Review by Casper64 — APPROVED** [review]

File: vlib/vweb/vweb.v [hunk 1]
```diff
@@ -620,6 +622,8 @@ fn handle_conn[T](mut conn net.TcpConn, global_app &T, routes &map[string]Route,
 		return
 	}
 
+	host := req.header.get(http.CommonHeader.host) or { '' }.to_lower()
```
[the remark this answers is not in the document]
Comment (Lenni0451) [reply vlib/vweb/vweb.v:625]: When trying to use `req.url.hostname()` I get an error when compiling: `error: unknown method or field: 'string.hostname'`

[the remark this answers is not in the document]
Comment (Casper64) [reply vlib/vweb/vweb.v:625]: My bad, you can use the `URL` struct created on line 610 not on `req`: `url.hostname()`

[the remark this answers is not in the document]
Comment (Lenni0451) [reply vlib/vweb/vweb.v:625]: I just tested printing the url struct and it is completely empty. This is the struct when connecting to `http://localhost/testing`:
`
URL{
  scheme:
  opaque:
  user: &nil
  host:
  path: /testing
  raw_path:
  force_query: false
  raw_query:
  fragment:
}
`

[the remark this answers is not in the document]
Comment (Lenni0451) [reply vlib/vweb/vweb.v:625]: Actually, `req.url` is just `/testing` in that case

[the remark this answers is not in the document]
Comment (Casper64) [reply vlib/vweb/vweb.v:625]: Well I guess that's another bug! Will fix that later, your implementation works for now

**Pull request merged**